File: cpp/cpp_100-999/478.cc
// C++
// leetcode 478
// https://leetcode.cn/problems/generate-random-point-in-a-circle/

class Solution {
public:
    double get() {
        return double(rand()) / RAND_MAX * 2 - 1;
    }

    double x, y, r;
    double PI = acos(-1);
    Solution(double radius, double x_center, double y_center) {
        r = radius;
        x = x_center;
        y = y_center;
    }
    
    vector<double> randPoint() {
        double a = get(), b = get();
        while(a * a + b * b > 1.0) {
            a = get();
            b = get();
        }
        return vector<double>{x + r * a, y + r * b};
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(radius, x_center, y_center);
 * vector<double> param_1 = obj->randPoint();
 */

File: Go/Go_100-999/478.go
// Go
// leetcode 478
// https://leetcode.cn/problems/generate-random-point-in-a-circle/

type Solution struct {
    x, y, r float64
}


func Constructor(radius float64, x_center float64, y_center float64) Solution {
    return Solution{x_center, y_center, radius}
}


func (this *Solution) RandPoint() []float64 {
    rr := rand.Float64() * this.r
    angle := rand.Float64() * this.r
    if angle > rr {
        rr = this.r - rr
        angle = (rr + angle - this.r) / rr * 2 * math.Pi
    } else {
        angle = angle / rr * 2 * math.Pi
    }
    return []float64{this.x + math.Cos(angle) * rr, this.y + math.Sin(angle) * rr}
}


/**
 * Your Solution object will be instantiated and called as such:
 * obj := Constructor(radius, x_center, y_center);
 * param_1 := obj.RandPoint();
 */